### PR TITLE
Add some debug friendly utilities for dag request based test framework

### LINKED ISSUE #4304: Add more detail information when DAGRequest test throws exception
## Enhancement
We can at least provide nth request throws the exception.

### PR DESCRIPTION
**Pull request by yibin87**
Signed-off-by: yibin <huyibin@pingcap.com>

### What problem does this PR solve?

Issue Number: close #4304 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by windtalker — APPROVED** [review]
LGTM

File: dbms/src/Debug/dbgFuncCoprocessor.cpp [hunk 1]
```diff
@@ -339,56 +340,99 @@ void dbgFuncTiDBQueryFromNaturalDag(Context & context, const ASTs & args, DBGInv
...
+    kvrpcpb::Context req_context = req.context();
+    RegionID region_id = req_context.region_id();
+    tipb::DAGRequest dag_request = getDAGRequestFromStringWithRetry(req.data());
+    RegionPtr region = context.getTMTContext().getKVStore()->getRegion(region_id);
+    if (!region)
+        throw Exception(fmt::format("No such region: {}", region_id), ErrorCodes::BAD_ARGUMENTS);
+
+    bool unequal_flag = false;
+    DAGProperties properties = getDAGProperties("");
+    std::__1::vector<std::__1::pair<DecodedTiKVKeyPtr, DecodedTiKVKeyPtr>> key_ranges = CoprocessorHandler::GenCopKeyRange(req.ranges());
```
Comment (ywqzzy) [inline dbms/src/Debug/dbgFuncCoprocessor.cpp:411]: To satisfy my curiosity, Could you tell me why use the "std::__1" namespace?

File: dbms/src/Debug/dbgFuncCoprocessor.cpp [hunk 2]
```diff
@@ -339,56 +340,99 @@ void dbgFuncTiDBQueryFromNaturalDag(Context & context, const ASTs & args, DBGInv
     auto dag = NaturalDag(json_dag_path, &Poco::Logger::get("MockDAG"));
     dag.init();
     dag.build(context);
-    bool unequal_flag = false;
-    String unequal_msg;
+    std::vector<std::pair<int32_t, String>> failed_req_msg_vec;
```
Comment (ywqzzy) [inline dbms/src/Debug/dbgFuncCoprocessor.cpp:343]:
```suggestion
    std::vector<std::pair<Int32, String>> failed_req_msg_vec;
```
Maybe better?

Comment (yibin87) [reply dbms/src/Debug/dbgFuncCoprocessor.cpp:343]: They seem the same to me:) And both usages can be found in our projects.

File: dbms/src/Debug/dbgFuncCoprocessor.cpp (hunk 1 above)
Comment (yibin87) [reply dbms/src/Debug/dbgFuncCoprocessor.cpp:411]: Ahoh, they were introduced by CLion's extracting method functionality, I'll change it.

File: dbms/src/Debug/dbgFuncCoprocessor.cpp [hunk 3]
```diff
@@ -339,56 +340,99 @@ void dbgFuncTiDBQueryFromNaturalDag(Context & context, const ASTs & args, DBGInv
...
+            if (first)
+            {
+                merged_msg = msg;
+                first = false;
+            }
+            else
+            {
+                merged_msg = fmt::format("{}\n{}", merged_msg, msg);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (yibin87) [reply dbms/src/Debug/dbgFuncCoprocessor.cpp:395]: Nice sugguestion!

File: dbms/src/Debug/dbgFuncCoprocessor.cpp (hunk 1 above)
Comment (ywqzzy) [reply dbms/src/Debug/dbgFuncCoprocessor.cpp:411]: Get.

File: dbms/src/Debug/dbgFuncCoprocessor.cpp [hunk 4]
```diff
@@ -378,28 +378,29 @@ void dbgFuncTiDBQueryFromNaturalDag(Context & context, const ASTs & args, DBGInv
     if (!failed_req_msg_vec.empty())
     {
         output("Invalid");
-        String merged_msg;
+        FmtBuffer msg_buffer;
```
[the remark this answers is not in the document]
Comment (yibin87) [reply dbms/src/Debug/dbgFuncCoprocessor.cpp]: Yeah, that's exactly what I needed. Used joinStr to simplify this process.

**Review by ywqzzy — APPROVED** [review]
LGTM

**Pull request merged**